File: api/routers/materiais.py
from fastapi import APIRouter, Depends, Query

import db
from api.deps import usuario_atual

router = APIRouter(prefix="/materiais", tags=["materiais"])


@router.get("")
def listar(
    area_id: int | None = None, especialidade_id: int | None = None, subtopico_id: int | None = None,
    tipo: str | None = None, q: str | None = None,
    limite: int = Query(default=50, ge=1, le=200), pagina: int = Query(default=0, ge=0),
    usuario=Depends(usuario_atual),
):
    total = db.contar_materiais_filtrados(area_id, subtopico_id, tipo, q, especialidade_id=especialidade_id)
    itens = db.listar_materiais_paginado(
        area_id, subtopico_id, tipo, q, limite=limite, offset=pagina * limite, especialidade_id=especialidade_id,
    )
    return {"total": total, "itens": itens}


@router.get("/tipos")
def tipos(usuario=Depends(usuario_atual)):
    return db.listar_tipos_materiais()

File: frontend/src/lib/materiais.ts
import { keepPreviousData, useQuery } from "@tanstack/react-query";
import { api } from "./api";
import type { Material } from "./types";

export function useMateriais(filtros: {
  area_id?: number;
  especialidade_id?: number;
  subtopico_id?: number;
  tipo?: string;
  q?: string;
  pagina: number;
  limite: number;
}) {
  return useQuery({
    queryKey: ["materiais", filtros],
    queryFn: () => api.get<{ total: number; itens: Material[] }>("/materiais", filtros),
    // Ao trocar filtro ou página a tabela anterior continua na tela (esmaecida)
    // até a nova chegar, em vez de piscar o esqueleto a cada clique.
    placeholderData: keepPreviousData,
  });
}

export function useTiposMateriais() {
  return useQuery({ queryKey: ["materiais-tipos"], queryFn: () => api.get<string[]>("/materiais/tipos") });
}

export function useStatusSincronizacao() {
  return useQuery({
    queryKey: ["sincronizacao-status"],
    queryFn: () => api.get<{ ultima_sincronizacao: string | null }>("/sincronizacao"),
  });
}
